File: 2025-and-before/0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        
    
        HashMap<Integer,Integer>mp = new HashMap<>();
        int ans = 0,sum = 0,n = nums.length; 
        for(int i=0;i<n;i++){
            sum = (sum%k + nums[i]%k + k)%k;
            
            if(sum == 0)
                ans++;
            
            int curr = mp.getOrDefault(sum,0);
            ans+=curr;
            
            mp.put(sum,curr+1);
        }
        
        return ans;
        
    }
}

File: 2025-and-before/0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        
        unordered_map<int,int>mp;
        int pre = 0,ans =0;
        for(int&i : nums){
            
            pre+=i;
            pre%=k;
            pre+=k;
            pre%=k;
            
            
            if(pre == 0)
                ans++;
            
            ans+=mp[pre];
            
            mp[pre]++;
        }
        
        return ans;
    }
};